mycpp/gc_heap.h

```diff
@@ -1,285 +1,37 @@
-    // bytes, because the GC pointer forwarding requires 16 bytes.  If we
-    // allocated less than 16 the GC would overwrite the adjacent object when
-    // it went to forward the pointer.
-    assert(n >= 16);
-
-#if GC_EVERY_ALLOC
-    Collect();  // force collection to find problems early
-#endif
-
-    if (free_ + n <= limit_) {  // Common case: we have space for it.
-      return Bump(n);
-    }
-
-#if GC_STATS
-      // log("GC free_ %p,  from_space_ %p, space_size_ %d", free_, from_space_,
-      //    space_size_);
-#endif
-
-    Collect();  // Try to free some space.
-
-    // log("after GC: from begin %p, free_ %p,  n %d, limit_ %p",
-    //    from_space_.begin_, free_, n, limit_);
-
-    if (free_ + n <= limit_) {  // Now we have space for it.
-      return Bump(n);
-    }
-
-    // It's still too small.  Grow the heap.
-    int multiple = 2;
-    Collect((from_space_.size_ + n) * multiple);
-
-#if GC_STATS
-    num_forced_growths_++;
-#endif
-
-    return Bump(n);
-  }
-
-  void Swap() {
-    // Swap spaces for next collection.
-    char* tmp = from_space_.begin_;
-    from_space_.begin_ = to_space_.begin_;
-    to_space_.begin_ = tmp;
-
-    int tmp2 = from_space_.size_;
-    from_space_.size_ = to_space_.size_;
-    to_space_.size_ = tmp2;
-  }
-
-  void PushRoot(Obj** p) {
-    // log("PushRoot %d", roots_top_);
-    roots_[roots_top_++] = p;
-    // TODO: This should be like a malloc() failure?
-    assert(roots_top_ < kMaxRoots);
-  }
-
-  void PopRoot() {
-    roots_top_--;
-    // log("PopRoot %d", roots_top_);
-  }
-
-  Obj* Relocate(Obj* obj, Obj* header);
-
-  // mutates free_ and other variables
-  void Collect(int to_space_size = 0);
-
-#if GC_STATS
-  void Report() {
-    log("-----");
-    log("num collections = %d", num_collections_);
-    log("num heap growths = %d", num_heap_growths_);
-    log("num forced heap growths = %d", num_forced_growths_);
-    log("num live objects = %d", num_live_objs_);
-
-    log("from_space_ %p", from_space_.begin_);
-    log("to_space %p", to_space_.begin_);
-    log("-----");
-  }
-#endif
-
-  Space from_space_;  // space we allocate from
-  Space to_space_;    // space that the collector copies to
-
-  char* free_;   // next place to allocate, from_space_ <= free_ < limit_
-  char* limit_;  // end of space we're allocating from
+#define Terabytes(bytes) (Gigabytes(bytes) * 1024)
+#define Gigabytes(bytes) (Megabytes(bytes) * 1024)
+#define Megabytes(bytes) (Kilobytes(bytes) * 1024)
+#define Kilobytes(bytes) ((bytes)*1024)
 
-  // Stack roots.  The obvious data structure is a linked list, but an array
-  // has better locality.
-  //
-  // femtolisp uses a global pointer to dynamically-allocated growable array,
-  // with initial N_STACK = 262144!  Kind of arbitrary.
+const int kMaxRoots = Kilobytes(4);
 
-  int roots_top_;
-  Obj** roots_[kMaxRoots];  // These are pointers to Obj* pointers
+#include "cheney_heap.h"
+#include "marksweep_heap.h"
```

Comment:
Please put all the #includes at the top of the file (or delete if they don't need to be here? )  If it's just for the `extern`, there is no #include needed

I expanded this section:

https://github.com/oilshell/oil/wiki/Oil-Ninja-Build#include-style

Reply:
Fixed.